Change canvas/context/snapshot size from u64 -> u32

hashcatHitman opened the pull request:
Replaces uses of `euclid::default::Size2D<u64>` with `euclid::default::Size2D<u32>` for the canvas/context/snapshot.

This PR includes changes to the following components:
 - canvas
 - pixels
 - script
 - script_bindings
 - shared/canvas
 - shared/snapshot

Testing: https://github.com/hashcatHitman/servo/actions/runs/15426115391 (as of 892edc0048dfad28342e5e2ff247c963eb91ed11)

Fixes: #36706

hashcatHitman committed:
Initial move of canvas/context/snapshot size from u64 -> u32

Changed the two instances of `euclid::default::Size2D<u64>` in
`servo/components/script/canvas_context.rs` and
`servo/components/shared/snapshot/lib.rs` that were outlined as the bare minimum
in "Make canvas/context/snapshot size be u32 everywhere" to
`euclid::default::Size2D<u32>`. Every other change made in this commit is
either:
 - of similar nature, and is the minimum that would allow compilation
 - resolving lints triggered by the former

More might be needed to complete the issue, but I feel like this is a good
starting point.

This commit includes changes to the following components:
 - canvas
 - pixels
 - script
 - shared/canvas
 - shared/snapshot

Signed-off-by: hashcatHitman <155700084+hashcatHitman@users.noreply.github.com>

sebsebmc commented:
You can also `./mach try wpt -r [name of your git fork remote]` to run the test runner in your GitHub repo.

hashcatHitman commented:
> You can also `./mach try wpt -r [name of your git fork remote]` to run the test runner in your GitHub repo.

Ah, thanks! I was wondering how that worked.

sagudev commented on components/script_bindings/webidls/OffscreenCanvas.webidl:
```diff
@@ -15,9 +15,9 @@ dictionary ImageEncodeOptions {
 
 [Exposed=(Window,Worker)/*, Transferable*/, Pref="dom_offscreen_canvas_enabled"]
 interface OffscreenCanvas : EventTarget {
-  [Throws] constructor([EnforceRange] unsigned long long width, [EnforceRange] unsigned long long height);
-  attribute [EnforceRange] unsigned long long width;
-  attribute [EnforceRange] unsigned long long height;
```
webidl should not be changed, because we want to still match the spec: https://html.spec.whatwg.org/multipage/canvas.html#the-offscreencanvas-interface

If user provided too big width or height, we say that the context is invalid (inpaintable), that's why in the internals we can deal with u32, because that's effective usable canvas size.

hashcatHitman replied:
Ah, I took a look at the spec but didn't register that this was a breaking change. I've resolved this in the latest revision, though I've introduced panics, as I'm assuming I shouldn't be changing the other signatures to be fallible.

Notably, the constructor already is fallible, but I'm not sure what `DOMException` would be appropriate, if any, so I've gone with a panic for now to better centralize how we handle this situation.

sagudev commented on components/canvas/canvas_paint_thread.rs:
```diff
@@ -184,7 +184,7 @@ impl<'a> CanvasPaintThread<'a> {
             Canvas2dMsg::DrawEmptyImage(image_size, dest_rect, source_rect) => {
```
Here we should change type of `image_size` to contain u32.

hashcatHitman replied:
Just to clarify - are you saying `Canvas2dMsg::DrawEmptyImage(Size2D<f64>, Rect<f64>, Rect<f64>)` should be `Canvas2dMsg::DrawEmptyImage(Size2D<u32>, Rect<u32>, Rect<u32>)`<sup>1</sup>? I knew we were going from `u64` to `u32`, but I was under the impression these wouldn't be floats to begin with unless at least one of the following were true:

- it is required by the specification somewhere
- we need the capability to specify floating point numbers

1: Changing the `Rect<f64>`s to `Rect<u32>`s is implied by your other comment of similar nature and the underlying relationship with `canvas::canvas_paint_thread::Canvas::draw_image`, which itself would presumably need to be changed as well.

sagudev replied:
I needed some time to remember why I said this, but I think I figured it out again. Spec only provides doubles: sx, sy, sw, sh, dx, dy, dw, dh that specify src and dest rectangles, so we should keep them as `Rect<f64>`. But I think we can safely change `image_size` to `Size2D<u32>`, so the final type should be `Canvas2dMsg::DrawEmptyImage(Size2D<u32>, Rect<f64>, Rect<f64>)`

hashcatHitman replied:
This, the [other instance], and the [panic] should all be resolved as of 892edc0048dfad28342e5e2ff247c963eb91ed11.

[other instance]: https://github.com/servo/servo/pull/36827#discussion_r2076736510
[panic]: https://github.com/servo/servo/pull/36827#discussion_r2076763261

sagudev commented on components/canvas/canvas_paint_thread.rs:
```diff
@@ -200,10 +200,10 @@ impl<'a> CanvasPaintThread<'a> {
             ) => {
```
Here we should change type of `image_size` and `rect` to contain u32.

hashcatHitman replied:
Similarly to my [other comment], but `DrawImageInOther(CanvasId, Size2D<f64>, Rect<f64>, Rect<f64>, bool)` to `DrawImageInOther(CanvasId, Size2D<u32>, Rect<u32>, Rect<u32>, bool)`?

[other comment]: https://github.com/servo/servo/pull/36827#discussion_r2119443308

sagudev commented on components/script/dom/offscreencanvas.rs:
```diff
@@ -116,6 +116,25 @@ impl OffscreenCanvas {
         self.placeholder.as_deref()
     }
+
+    /// Tries to convert a [`u64`] into [`u32`] for the purposes of
+    /// compatibility with the [OffscreenCanvas spec].
+    ///
+    /// # Panics
+    ///
+    /// Panics if converting the provided [`u64`] into a [`u32`] with
+    /// [`u64::try_into`] would return [`Result::Err`].
+    ///
+    /// [OffscreenCanvas spec]: https://html.spec.whatwg.org/multipage/#the-offscreencanvas-interface
+    fn try_fit(value: u64) -> u32 {
+        match value.try_into() {
+            Ok(fits) => fits,
+            Err(error) => {
+                let msg = format!("{error}\n\nInvalid canvas dimension \"{value}\"!");
+                panic!("{}", msg)
+            },
+        }
+    }
```
We should not panic. Spec is really under specified here. But firefox throws RangeError if too big (in constructor or setters) while chromium restrict constructor to unsigned long but allows setting any value in setters.

What I think we should do is store u64 (for constructor and setters) and convert u64 to u32 (by clamping any too big numbers to `u32::MAX`) when obtaining size for context (this is `get_size`) thus effectively letting actual context (2d, webgl, webgpu) handle too big size.

hashcatHitman replied:
This sounds good to me. I don't like to panic, but I wasn't really sure what the correct answer was here.

hashcatHitman commented:
In case anyone is wondering: I have not abandoned this. Final projects/exams have just had me a bit busy lately, but I will be working on this again as soon as I can.

sagudev commented:
No worries. Good luck with the exams.

hashcatHitman committed:
Size2D<f64> -> Size2D<u32> in canvas_traits::canvas::Canvas2dMsg

Changed the variants `canvas_traits::canvas::Canvas2dMsg::DrawEmptyImage` and
`canvas_traits::canvas::Canvas2dMsg::DrawImageInOther` to use
`euclid::default::Size2D<u32>` instead of `euclid::default::Size2D<f64>`. Every
other change made in this commit is either:
 - of similar nature, and is the minimum that would allow compilation
 - resolving lints triggered by the former

This commit includes changes to the following components:
 - canvas
 - script
 - shared/canvas

Signed-off-by: hashcatHitman <155700084+hashcatHitman@users.noreply.github.com>

sagudev reviewed (commented):
LGTM

did you do wpt run? is there any reason for this PR to be draft/WIP?

sagudev commented:
> I did `./mach try wpt -r [my-remote]`, the results of which are [here](https://github.com/hashcatHitman/servo/actions/runs/15426115391) (also linked in the PR description).

Whoops, missed that.

> I don't remember if this is the project I was looking at that gave contributors the choice of squashing/merging or not, but if it is, I'll note I generally intend for my commits to be applied with a clean rebase whenever possible. If not, just do whatever you usually do, I'm thinking of a different project.

We use merge queue that does rebase&CI then squash after all CI passes, following [Not Rocket Science Rule Of Software Engineering](https://gordonbrander.com/pattern/not-rocket-science-rule/).

sagudev reviewed (approved):
Great first contribution!
